Search result items for urgent care search mashup

brangi opened the pull request:
## Description
Issue:https://github.com/department-of-veterans-affairs/va.gov-team/issues/11460

## Testing done
Locally with mock data

## Screenshots

**Mashup Results** (mock data)
<img width="1276" alt="Screen Shot 2020-08-10 at 9 28 41 PM" src="https://user-images.githubusercontent.com/100468/89850801-9ca2c100-db50-11ea-8693-80d0fad71b48.png">



## Acceptance criteria
- [x] Veteran can choose to view urgent care results as 
      - VA urgent care locations **and** Community urgent care providers (in VA's network) in a single list
      - VA urgent care locations **only**
      - Community urgent care providers (in VA's network) **only**
- [x] When VA facility type = Urgent care and VA service type drop down = All urgent care, the search results list should include **both** VA urgent care locations and Community urgent care providers (in VA's network), organized in ascending order by distance 
- [x] When VA facility type = Urgent care and VA service type drop down = All urgent care, the search results list should include **only** VA urgent care locations will contain a facility operating status alert box (if appropriate based on data) 
- [x] When VA facility type = Urgent care and VA service type drop down = All urgent care, the search results list should include **only**  Community urgent care providers (in VA's network) will contain a  hyperlinked "In-network urgent care benefit" info alert box.  
- [x] When VA facility type = Urgent care and VA service type drop down = VA urgent care, the search results list should include **only** VA urgent care locations organized in ascending order by distance  
- [x] When VA facility type = Urgent care and VA service type drop down = Community urgent care providers (in VA's network), the search results list should include **only** Community urgent care providers (in VA's network) organized in ascending order by distance 
   

## Definition of done
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
Init Search results card design and behavior for urgent care mash-up

Commit:
Start refactor components

brangi commented on src/applications/facility-locator/components/search-results-items/VaFacilityResult.jsx:
```diff
@@ -0,0 +1,95 @@
+    <div
+      className={`usa-alert ${classNameAlert} background-color-only notice-marg-pad`}
+    >
+      <i
```
Nothing has changed

Commit:
Refactor more

brangi committed:
Refactor call search cases

brangi committed:
fix test case

mpelzsherman commented:
Still digging into this, but I'm seeing some odd behavior on this branch.

Search for Urgent Care, Service Type "VA Urgent Care" in 27606.

The API returns one record (name: "Fayetteville VA Medical Center"), but nothing is displayed in the search results.

I can't reproduce this on staging so I think it's a bug in your branch. I'll see if I can get you more info.

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +41,62 @@ class ResultsList extends Component {
+            // New option
+            // All mash up
+            item = <UrgentCareResult provider={r} query={query} />;
+          }
```
Unless the user selects a service type, `query.serviceType` is null, and nothing is rendered here.

mpelzsherman replied:
We may want to set a default serviceType in the searchQuery reducer.

brangi committed:
Fix case all the va facilities

brangi committed:
Add mock urgent care mashup test data

brangi committed:
Update the mock localhost API locator

brangi committed:
Fix urgent care and service null case
Fix MockApiLocator to work with localhost

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +41,67 @@ class ResultsList extends Component {
+            item = <UrgentCareResult provider={r} query={query} />;
+          } else if (!query.serviceType || query.serviceType === 'UrgentCare') {
+            // this is probably going to change, it should default to All,
+            // currently the logic is if serviceType null, the result is VA urgent care
```
This works, but it's kind of fragile. I would prefer to set the default serviceType in the searchQuery reducer. The UI should match the query.

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -16,6 +16,10 @@ import { updateSearchQuery, searchWithBounds } from '../actions';
 
 import SearchResult from './SearchResult';
```
We should delete this since SearchResult isn't being used.

brangi committed:
More mock data to localhost locator

mpelzsherman commented:
Love the refactoring work here!

I'd like to see unit tests for each of the new sub-components (CCProviderResult, PharmacyResult, etc.)

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/UrgentCareResult.jsx:
```diff
@@ -0,0 +1,102 @@
+import { ccUrgentCareLabels } from '../../config';
+import { CLINIC_URGENTCARE_SERVICE, LocationType } from '../../constants';
+
+export const urgentCareCall = query => {
```
I think it's best practice to encapsulate these functions inside the class.

mpelzsherman commented on src/applications/facility-locator/tests/components/UrgentCareSearch.unit.spec.js:
```diff
@@ -1,8 +1,10 @@
+import {
+  urgentCareCall,
+  posProviderName,
+} from '../../components/search-results-items/UrgentCareResult';
```
I'd suggest we move the tests for `urgentCareCall` and `posProviderName`  into an `UrgentCareResult` unit test.

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/VaFacilityResult.jsx:
```diff
@@ -0,0 +1,106 @@
+import { OperatingStatus } from '../../constants';
+import LocationAddress from './common/LocationAddress';
+
+// This could probably be its own component ?
```
Yes, or at least move the function inside the class.

brangi committed:
Merge branch 'master' into fl-issue-mashup-part2

Commit:
Merge branch 'master' into fl-issue-mashup-part2

Commit:
Merge branch 'fl-issue-mashup-part2' of github.com:department-of-veterans-affairs/vets-website into fl-issue-mashup-part2

Commit:
Fixes

U-DON commented on src/applications/facility-locator/components/search-results-items/common/LocationOperationStatus.jsx:
```diff
@@ -0,0 +1,40 @@
+    <div
+      className={`usa-alert ${classNameAlert} background-color-only notice-marg-pad`}
+    >
+      <i
```
Seems presentational, so needs `aria-hidden="true"`.

brangi replied:
you mean the icon ?

U-DON replied:
Yup, the `<i>` element.

Commit:
Merge branch 'master' into fl-issue-mashup-part2

Commit:
Fixes to test cases

Commit:
More progress on refactoring

brangi committed:
Merge branch 'master' into fl-issue-mashup-part2

brangi committed:
Update test cases snapshots

brangi committed:
Fixes and clean up

brangi committed:
Fixes

brangi committed:
Add Info alert benefit link and fixes

brangi committed:
Add unit tests for new components

brangi committed:
Add rest of unit test cases

brangi committed:
Add more test coverage

brangi committed:
Design clean up

mpelzsherman commented:
> The API returns one record (name: "Fayetteville VA Medical Center"), but nothing is displayed in the search results.

This is fixed in the current review env.

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/CCProviderResult.jsx:
```diff
@@ -0,0 +1,52 @@
+ */
+const CCProviderResult = ({ provider, query }) => {
+  const { name } = provider.attributes;
+  const distance = provider.distance;
```
Do we need to guard against provider or provider.attributes being null here?

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/UrgentCareResult.jsx:
```diff
@@ -0,0 +1,116 @@
+    ) {
+      return <p>Call to confirm services and hours</p>;
+    }
+    return null;
```
Can we add a line break above this line? I think it helps w/readability, otherwise it looks like this line is grouped with the previous block.

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/UrgentCareResult.jsx:
```diff
@@ -0,0 +1,116 @@
+          from={'SearchResult'}
+          query={query}
+        />
+        {urgentCareCall(provider, query)}
```
I don't think we need to pass `provider` and `query` as arguments here since they are in scope for `urgentCareCall`. Using them directly avoids confusion stemming from having to re-name the variables to avoid shadowing.

mpelzsherman commented on src/applications/facility-locator/tests/components/UrgentCareSearch.unit.spec.jsx:
```diff
@@ -0,0 +1,30 @@
+    expect(wrapper).to.matchSnapshot();
+    wrapper.unmount();
+  });
+  it('Should render UrgentCareResult VA', () => {
```
Please add an empty line between `it` blocks, thanks!

mpelzsherman commented on src/applications/facility-locator/tests/components/VaFacilitySearch.unit.spec.jsx:
```diff
@@ -0,0 +1,52 @@
+import testData from '../../constants/mock-facility-data-v1.json';
+import { LocationType } from '../../constants';
+
+describe('VaFacilitySearch', () => {
```
The name of the test (and its file) should match the component under test, so I think we should change `VaFacilitySearch`to `VaFacilityResult` here and in the filename.

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +44,66 @@ class ResultsList extends Component {
+          item = <PharmacyResult provider={r} query={query} />;
+          break;
+        case 'urgent_care':
+          if (query.serviceType === 'NonVAUrgentCare') {
+            item = <UrgentCareResult provider={r} query={query} />;
+          } else if (query.serviceType === 'UrgentCare') {
+            item = <VaFacilityResult location={r} query={query} />;
+          } else if (
+            !query.serviceType ||
+            query.serviceType === 'AllUrgentCare'
+          ) {
+            // New option
+            // All mash up
+            if (r.type === LocationType.CC_PROVIDER) {
+              item = <UrgentCareResult provider={r} query={query} />;
+            } else {
+              item = <VaFacilityResult location={r} query={query} />;
+            }
```
What do you think about simplifying this logic like so:

```suggestion
            if (r.type === LocationType.CC_PROVIDER) {
              item = <UrgentCareResult provider={r} query={query} />;
            } else {
              item = <VaFacilityResult location={r} query={query} />;
            }
```

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +44,66 @@ class ResultsList extends Component {
+          item
+        </div>
+      ) : (
+        item
```
Oops! You have a bug here:

```suggestion
        {item}
```

mpelzsherman replied:
I'm seeing a warning about the lack of a "key" prop here.

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +44,66 @@ class ResultsList extends Component {
+      }
+      item = isMobile ? (
+        <div key={r.id} className="mobile-search-result">
+          item
```
Oops! You have a bug here:

```suggestion
        {item}
```

mpelzsherman reviewed (changes requested):
Please see comments, esp. the bug with mobile results.

Commit:
Test cases fixes and clean up

Commit:
Fixes in logic and bugs

Commit:
add mashup new endpoint

Commit:
Fix test case

mpelzsherman commented:
Looks like you resolved the more serious issues. My remaining comments are optional.

The review env isn’t working though - I’m seeing 404 error from the API calls. And you have a merge conflict.

brangi committed:
Parse included object for mashup search

brangi committed:
Merge branch 'master' into fl-issue-mashup-part2

brangi committed:
clean up old alert to urgent care benfit link

brangi committed:
Add more mock data

brangi committed:
more clean up

mpelzsherman commented on src/applications/facility-locator/components/ResultsList.jsx:
```diff
@@ -37,6 +44,55 @@ class ResultsList extends Component {
+          item = null;
+      }
+      item = isMobile ? (
+        <div key={r.id} className="mobile-search-result">
```
Let's check with Leyda, but I don't think this class is needed, and removing it would make the `isMobile` param unnecessary (better for responsive design).

```css
  .mobile-search-result {
    padding: 1em 0;    // doesn't seem to be needed
    border-top: none;  // has no effect
  }
```

brangi replied:
let's add this kind of changes to the design review

Commit:
Handle mashup api response

mpelzsherman committed:
Design changes:

- Move Directions link closer to address and add space below
- Convert all-caps PPMS data to Title Case

brangi committed:
Update snapshots

jbalboni commented on src/applications/facility-locator/utils/facilityAddress.js:
```diff
@@ -1,15 +1,16 @@
 import { compact, isEmpty } from 'lodash';
 import { LocationType } from '../constants';
+import { titleCase } from 'applications/vaos/utils/formatters';
```
It would be better for you to copy this code into the facility-locator code, or if you think it's broadly useful, add it somewhere in the platform folder, instead of using it from vaos directly. Keeping applications as independent as we can helps avoid teams accidentally breaking other teams code without knowing it and also allows us to refactor our own apps without needing to change someone else's.

mpelzsherman replied:
Yeah, I was wondering about that.

I did create a Facility-specific version of the `titleCase` function, but it has some domain-specific behavior that we don't need in all cases. I'll modify our function to be more flexible, and use it instead.

mpelzsherman committed:
Move address formatting code under facilities, add tests

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/VaFacilityResult.jsx:
```diff
@@ -0,0 +1,70 @@
+ * VA benefits
+ * Vet Centers
+ * VA cemeteries
+ * VA Urgent care
```
Why is this comment here?

mpelzsherman commented on src/applications/facility-locator/components/search-results-items/common/LocationOperationStatus.jsx:
```diff
@@ -0,0 +1,41 @@
+  let infoMsg;
+  let classNameAlert;
+  let iconType;
+  if (operatingStatus.code === OperatingStatus.NOTICE) {
```
Can we please add an empty line above each `if` block and above the `return` statement?

mpelzsherman committed:
remove title-casing of PPMS data per Dave Conlon

mpelzsherman committed:
Merge branch 'master' into fl-issue-mashup-part2

mpelzsherman committed:
update snapshots

mpelzsherman committed:
minor code cleanup

mpelzsherman committed:
Merge branch 'master' into fl-issue-mashup-part2

mpelzsherman committed:
css changes

- border top on result cards (mobile & desktop)
- made result list wider

mpelzsherman committed:
css changes

- search form labels
- map pin click card highlight color
- icon/text spacing in CTA's
- fix icon for "In network urgent care benefit" CTA

Commit:
fix - remove urgent care link from ccp results

Commit:
Fix test cases

Commit:
Fix call to action missing on pharmacy result and tc fixes

mpelzsherman committed:
Send service type to analytics

Commit:
fix for markers onclick action

Commit:
Merge branch 'master' into fl-issue-mashup-part2

mpelzsherman committed:
fix lint errors

Commit:
Fix - autolocation enabled should only preset the input address and not perform search

Commit:
Fix - update search query with place name

mpelzsherman committed:
code cleanup and message tweak

mpelzsherman committed:
text cleanup

mpelzsherman committed:
Show default service type when "All" selected

mpelzsherman committed:
Merge branch 'master' into fl-issue-mashup-part2
